bson/bsoncodec/array_codec.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncodec
+
+import (
+	"reflect"
+
+	"go.mongodb.org/mongo-driver/bson/bsonrw"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+// ArrayCodec is the Codec used for bsoncore.Array values.
+type ArrayCodec struct{}
+
+var defaultArrayCodec = NewArrayCodec()
+
+// NewArrayCodec returns an ArrayCodec.
+func NewArrayCodec() *ArrayCodec {
+	return &ArrayCodec{}
+}
+
+// EncodeValue is the ValueEncoder for bsoncore.Array values.
+func (ac *ArrayCodec) EncodeValue(ec EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
+	if !val.IsValid() || val.Type() != tCoreArray {
+		return ValueEncoderError{Name: "CoreArrayEncodeValue", Types: []reflect.Type{tCoreArray}, Received: val}
+	}
+
+	arr := val.Interface().(bsoncore.Array)
+	return bsonrw.Copier{}.CopyArrayFromBytes(vw, arr)
+}
+
+// DecodeValue is the ValueDecoder for bsoncore.Array values.
+func (ac *ArrayCodec) DecodeValue(dc DecodeContext, vr bsonrw.ValueReader, val reflect.Value) error {
+	if !val.CanSet() || val.Type() != tCoreArray {
+		return ValueDecoderError{Name: "CoreArrayDecodeValue", Types: []reflect.Type{tCoreArray}, Received: val}
+	}
+
+	if val.IsNil() {
+		val.Set(reflect.MakeSlice(val.Type(), 0, 0))
+	}
+
+	val.SetLen(0)
+	_, arr, err := bsonrw.Copier{}.AppendValueBytes(val.Interface().(bsoncore.Array), vr)
```

Comment:
The decoder for `bsoncore.Document` uses `Copier#AppendDocumentBytes`, which internally calls `Copier#CopyDocument`. The latter function knows that we're working with documents, so it can call `ReadDocument` to get a `DocumentReader` and pass that down to `Copier#copyDocumentCore`.

Should we be adding similar functions for arrays rather than using the generic `Copier#AppendValueBytes`? If so, I think we can add similar logic to pass down functions to `copyDocumentCore` that abstract away the differences between `DocumentReader` and `ArrayReader`.

Reply:
Good point. I think it would be good to have similar logic for `bsoncore.Array`, so I've added `Copier#AppendArrayBytes()` to call from `array_codec#DecodeValue()`. It's the exact same as `Copier#AppendDocumentBytes()`, but it uses the existing `Copier#copyArray()`. So, no need to use `copyDocumentCore()` I think.